Add Vacuum.RemoveLayerLink function

waynr opened the pull request:
Currently MarkAndSweep garbage collection only deletes manifests, manifest tag references, and blobs.

It leaves behind orphaned "link" files that look like `<root>/v2/repositories/<name>/_layers/<algorithm>/<hex digest>/link`; these are repository-specific links to items in the blob store that we no longer need once the blob has been deleted.

adamwg commented on registry/storage/garbagecollect.go:
```diff
@@ -122,6 +139,14 @@ func MarkAndSweep(ctx context.Context, storageDriver driver.StorageDriver, regis
 			}
+			for _, layerDgst := range obj.Layers {
+				if _, ok := markSet[layerDgst]; !ok {
+					err := vacuum.RemoveLayerLink(obj.Name, string(layerDgst))
```
I see we're casting the `layerDgst` to a string here, then parsing it back into a `digest.Digest` in the vacuum. Could we pass it as a `digets.Digest` to avoid the round-trip?

adamwg commented on registry/storage/vacuum.go:
```diff
@@ -84,6 +84,38 @@ func (v Vacuum) RemoveManifest(name string, dgst digest.Digest, tags []string) e
+		}
+	}
+
+	err = v.driver.Delete(v.ctx, layerLinkPath)
```
Nit: Could `return v.driver.Delete(...)` since we're not doing anything with the error other than returning it.

waynr commented:
Tests are currently failing with:

```
--- FAIL: TestLinkedBlobStoreCreateWithMountFrom (0.33s)
    linkedblobstore_test.go:135: expected entry "nm/foo@sha256:e06fcd9fa20ae3b3b53df53b9ada3e289cfe5c0f437790d7bd8f34acc9e11641" not found among handled stat calls
    linkedblobstore_test.go:135: expected entry "nm/foo@sha256:05f2bcb0267ea2d81cd684392fb2e9faf1be7006d6c9694c9f07f0f4f759e133" not found among handled stat calls
--- FAIL: TestManifestStorage (0.29s)
    manifeststore_test.go:320: Deleted manifest should not exist
    manifeststore_test.go:325: Unexpected success getting deleted manifest
panic: reflect: call of reflect.Value.Type on zero Value [recovered]
        panic: reflect: call of reflect.Value.Type on zero Value

goroutine 23 [running]:
testing.tRunner.func1.1(0x87b4e0, 0xc0001f6ca0)
        /home/wayne/usr/local/go/src/testing/testing.go:988 +0x30d
testing.tRunner.func1(0xc000389b00)
        /home/wayne/usr/local/go/src/testing/testing.go:991 +0x3f9
panic(0x87b4e0, 0xc0001f6ca0)
        /home/wayne/usr/local/go/src/runtime/panic.go:969 +0x166
reflect.Value.Type(0x0, 0x0, 0x0, 0x0, 0x0)
        /home/wayne/usr/local/go/src/reflect/value.go:1879 +0x183
github.com/docker/distribution/registry/storage.testManifestStorage(0xc000389b00, 0xc00038a701, 0xc00024af48, 0x5, 0x5)
        /home/wayne/projects/distribution/registry/storage/manifeststore_test.go:331 +0x244c
github.com/docker/distribution/registry/storage.TestManifestStorage(0xc000389b00)
        /home/wayne/projects/distribution/registry/storage/manifeststore_test.go:62 +0x1dc
testing.tRunner(0xc000389b00, 0x926970)
        /home/wayne/usr/local/go/src/testing/testing.go:1039 +0xdc
created by testing.(*T).Run
        /home/wayne/usr/local/go/src/testing/testing.go:1090 +0x372
FAIL    github.com/docker/distribution/registry/storage 11.009s
FAIL
```

which a git bisect identified as being due to https://github.com/digitalocean/docker-distribution/commit/5132136516826856edb290d59dfc74d191df9980, which is funny because that's the change I reverted in #5 and which @adamwg rereverted in #9; rerereverting it is currently blocked on:

* first ensuring that manifest creation results in layer links where possible
* then backfilling all the missing layer links in DOCR

Since we've tolerated this test failure unknowingly since https://github.com/digitalocean/docker-distribution/commit/5132136516826856edb290d59dfc74d191df9980 I am going to assume we don't need to fix it before merging this PR which is desirable for the garbage collection implementation I am working on.
